File: scripts/e2e/_lib/envelope.py
# Copyright (c) 2025 Beijing Volcano Engine Technology Co., Ltd.
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
# http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.

"""Envelope 与退出码的契约校验工具。"""

from __future__ import annotations

from typing import Any, Iterable, Mapping

VALID_STATUS = {"success", "failed"}
VALID_ERROR_KINDS = {
    "unknown",
    "auth_failed",
    "config_missing",
    "resource_not_found",
    "permission_denied",
    "validation_error",
    "rate_limited",
    "transfer_failed",
    "conflict",
}

EXIT_CODES = {
    "success": 0,
    "unknown": 1,
    "auth_failed": 2,
    "config_missing": 3,
    "resource_not_found": 4,
    "permission_denied": 5,
    "validation_error": 6,
    "rate_limited": 7,
    "transfer_failed": 8,
    "conflict": 9,
}


def assert_success_envelope(envelope: Mapping[str, Any], *, command_substr: str = "") -> None:
    """成功响应必须满足：status=success、command 非空、error 缺失或为 None。"""
    assert envelope.get("status") == "success", (
        f"expected status=success, got {envelope.get('status')!r}; envelope={envelope}"
    )
    cmd = envelope.get("command")
    assert isinstance(cmd, str) and cmd, f"expected non-empty command, got {cmd!r}"
    if command_substr:
        assert command_substr in cmd, f"expected '{command_substr}' in command, got {cmd!r}"
    err = envelope.get("error")
    assert err in (None, {}), f"success envelope must not carry error, got {err!r}"


def assert_failure_envelope(
    envelope: Mapping[str, Any],
    *,
    expected_kind: str,
) -> None:
    """失败响应必须包含结构化 error，且 kind 与预期一致。"""
    assert expected_kind in VALID_ERROR_KINDS, f"unknown error kind: {expected_kind}"
    assert envelope.get("status") == "failed", (
        f"expected status=failed, got {envelope.get('status')!r}; envelope={envelope}"
    )
    err = envelope.get("error")
    assert isinstance(err, dict), f"expected error dict, got {err!r}"
    assert err.get("kind") == expected_kind, (
        f"expected error.kind={expected_kind}, got {err.get('kind')!r}"
    )
    assert isinstance(err.get("message"), str) and err["message"], "error.message must be non-empty"
    assert isinstance(err.get("exit_code"), int), "error.exit_code must be int"
    assert err["exit_code"] == EXIT_CODES[expected_kind], (
        f"error.exit_code mismatch for kind={expected_kind}: "
        f"expected {EXIT_CODES[expected_kind]}, got {err['exit_code']}"
    )


def assert_exit_code(actual: int, expected_kind: str) -> None:
    expected = EXIT_CODES[expected_kind]
    assert actual == expected, f"expected exit code {expected} ({expected_kind}), got {actual}"


def envelope_data_iter(envelope: Mapping[str, Any]) -> Iterable[Any]:
    """统一 data 提取：兼容 dict / list / scalar。"""
    data = envelope.get("data")
    if data is None:
        return iter(())
    if isinstance(data, list):
        return iter(data)
    return iter([data])

File: crates/tos-core/src/agent/envelope.rs
/*
 * Copyright (c) 2025 Beijing Volcano Engine Technology Co., Ltd.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

use serde::{Deserialize, Serialize};

use super::error::AgentErrorCategory;

#[derive(Debug, Serialize, Deserialize)]
pub struct Envelope<T: Serialize> {
    pub success: bool,
    pub status: Status,
    pub command: String,
    pub request_id: Option<String>,
    pub status_code: Option<u16>,
    pub ec: Option<String>,
    pub data: Option<T>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub pagination: Option<PaginationInfo>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub error: Option<ErrorDetail>,
}

#[derive(Debug, Serialize, Deserialize)]
#[serde(rename_all = "snake_case")]
pub enum Status {
    Success,
    Failed,
}

#[derive(Debug, Serialize, Deserialize)]
pub struct PaginationInfo {
    /// Token-based pagination cursor used by TOS APIs.
    #[serde(skip_serializing_if = "Option::is_none")]
    pub next_token: Option<String>,
    /// Marker-based pagination cursor used by ADrive/IDS APIs.
    #[serde(skip_serializing_if = "Option::is_none")]
    pub next_marker: Option<String>,
    /// Number of entries returned in the current response.
    pub total_returned: u64,
}

#[derive(Debug, Serialize, Deserialize)]
pub struct ErrorDetail {
    #[serde(skip_serializing, default)]
    pub status_code: Option<u16>,
    pub code: String,
    pub message: String,
    pub exit_code: i32,
    pub kind: ErrorKind,
    pub category: AgentErrorCategory,
    pub suggested_action: Option<String>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub fix_command: Option<String>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub doctor_hint: Option<String>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub docs_url: Option<String>,
}

#[derive(Debug, Serialize, Deserialize)]
#[serde(rename_all = "snake_case")]
pub enum ErrorKind {
    Unknown,
    AuthFailed,
    ConfigMissing,
    ResourceNotFound,
    PermissionDenied,
    ValidationError,
    RateLimited,
    TransferFailed,
    Conflict,
}

impl<T: Serialize> Envelope<T> {
    pub fn success(command: impl Into<String>, data: T) -> Self {
        Self {
            success: true,
            status: Status::Success,
            command: command.into(),
            // [Review Fix #3] Even direct Envelope serialization should carry
            // a stable request_id; service responses overwrite this with the
            // upstream x-tos-request-id via with_request_id().
            request_id: Some(ulid::Ulid::new().to_string()),
            status_code: None,
            ec: None,
            data: Some(data),
            pagination: None,
            error: None,
        }
    }

    pub fn with_request_id(mut self, id: impl Into<String>) -> Self {
        self.request_id = Some(id.into());
        self
    }

    /// Mark an envelope as having no single upstream request ID.
    ///
    /// Use this for client-side aggregate commands that may issue zero or many
    /// service requests, where a generated fallback ID would be misleading.
    pub fn without_request_id(mut self) -> Self {
        self.request_id = None;
        self
    }

    pub fn with_pagination(mut self, pagination: PaginationInfo) -> Self {
        self.pagination = Some(pagination);
        self
    }
}

impl Envelope<()> {
    pub fn failed(command: impl Into<String>, error: ErrorDetail) -> Self {
        Self {
            success: false,
            status: Status::Failed,
            command: command.into(),
            request_id: Some(ulid::Ulid::new().to_string()),
            status_code: error.status_code,
            ec: Some(error.code.clone()),
            data: None,
            pagination: None,
            error: Some(error),
        }
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_success_envelope_serialization() {
        let env = Envelope::success("tos bucket list", serde_json::json!({"buckets": []}));
        let json = serde_json::to_string(&env).unwrap();
        let parsed: serde_json::Value = serde_json::from_str(&json).unwrap();
        assert_eq!(parsed["success"], true);
        assert_eq!(parsed["status"], "success");
        assert_eq!(parsed["command"], "tos bucket list");
        assert!(parsed["request_id"].is_string());
        assert!(parsed["status_code"].is_null());
        assert!(parsed["ec"].is_null());
        assert!(parsed["data"].is_object());
        assert!(parsed.get("error").is_none());
    }

    #[test]
    fn test_failed_envelope_serialization() {
        let error = ErrorDetail {
            status_code: Some(404),
            code: "NoSuchBucket".to_string(),
            message: "Bucket not found".to_string(),
            exit_code: 4,
            kind: ErrorKind::ResourceNotFound,
            category: AgentErrorCategory::NotFound,
            suggested_action: Some("verify the bucket name or create the bucket".to_string()),
            fix_command: Some("tos bucket create --bucket my-bucket".to_string()),
            doctor_hint: None,
            docs_url: None,
        };
        let env = Envelope::<()>::failed("tos object upload", error);
        let json = serde_json::to_string(&env).unwrap();
        let parsed: serde_json::Value = serde_json::from_str(&json).unwrap();
        assert_eq!(parsed["success"], false);
        assert_eq!(parsed["status"], "failed");
        assert_eq!(parsed["status_code"], 404);
        assert_eq!(parsed["ec"], "NoSuchBucket");
        assert_eq!(parsed["error"]["code"], "NoSuchBucket");
        assert_eq!(parsed["error"]["exit_code"], 4);
        assert_eq!(parsed["error"]["kind"], "resource_not_found");
    }

    #[test]
    fn test_envelope_with_pagination() {
        let env = Envelope::success("tos object list", serde_json::json!({"objects": []}))
            .with_pagination(PaginationInfo {
                next_token: Some("abc123".to_string()),
                next_marker: None,
                total_returned: 100,
            });
        let json = serde_json::to_string(&env).unwrap();
        let parsed: serde_json::Value = serde_json::from_str(&json).unwrap();
        assert_eq!(parsed["pagination"]["next_token"], "abc123");
        assert!(parsed["pagination"].get("next_marker").is_none());
        assert_eq!(parsed["pagination"]["total_returned"], 100);
    }

    #[test]
    fn test_envelope_with_marker_pagination() {
        let env = Envelope::success("adrive ls", serde_json::json!({"files": []})).with_pagination(
            PaginationInfo {
                next_token: None,
                next_marker: Some("marker-abc".to_string()),
                total_returned: 50,
            },
        );
        let json = serde_json::to_string(&env).unwrap();
        let parsed: serde_json::Value = serde_json::from_str(&json).unwrap();
        assert!(parsed["pagination"].get("next_token").is_none());
        assert_eq!(parsed["pagination"]["next_marker"], "marker-abc");
        assert_eq!(parsed["pagination"]["total_returned"], 50);
    }

    #[test]
    fn test_envelope_with_request_id() {
        let env = Envelope::success("test", serde_json::json!({})).with_request_id("req-12345");
        let json = serde_json::to_string(&env).unwrap();
        let parsed: serde_json::Value = serde_json::from_str(&json).unwrap();
        assert_eq!(parsed["request_id"], "req-12345");
    }
}
